File: 2021/java/554.brick-wall.java
class Solution {
    // hashmap, key: preSum, value: count 
    public int leastBricks(List<List<Integer>> wall) {
        if (wall == null || wall.size() == 0) return 0;
        Map<Integer, Integer> preSum = new HashMap<>();
        int max = 0;
        for (List<Integer> row : wall) {
            int sum = 0;
            for (int i = 0; i < row.size() - 1; i++) {
                sum += row.get(i);
                preSum.put(sum, preSum.getOrDefault(sum, 0) + 1);
                max = Math.max(max, preSum.get(sum));
            }
        }
        return wall.size() - max;
    }
}

File: 2021/python/554.brick-wall.py
class Solution:
    def leastBricks(self, wall: List[List[int]]) -> int:
        if not wall or len(wall) == 0:
            return 0
        pre_sum_map, max_sum = {}, 0
        for w in wall:
            pre_sum = 0;
            for c in range(len(w) - 1):
                pre_sum += w[c]
                pre_sum_map[pre_sum] = pre_sum_map.get(pre_sum, 0) + 1
                max_sum = max(max_sum, pre_sum_map.get(pre_sum))
        
        return len(wall) - max_sum
                
